keras/src/backend/jax/core.py

```diff
@@ -55,6 +58,248 @@ def __jax_array__(self):
+
+            def _direct_assign(self, value):
+                # Apply JAX-specific distribution if layout is present
+                if self._layout is not None:
+                    processed_value = distribution_lib.distribute_variable(
+                        value, self._layout
+                    )
+                else:
+                    processed_value = value
+
+                # Ensure that nnx.Variable part is initialized
+                if not hasattr(self, "_var_metadata"):
+                    # todo: should add a warning
+                    pass
+
+                # Apply on_set_value hook if it exists
+                if (
+                    hasattr(self, "_var_metadata")
+                    and "on_set_value" in self._var_metadata
+                ):
+                    final_value = self._var_metadata["on_set_value"](
+                        self, processed_value
+                    )
+                else:
+                    final_value = processed_value
+
+                # Directly set raw_value. nnx.Variable handles mutable array
+                # updates
+                object.__setattr__(self, "raw_value", final_value)
+
+            def _convert_to_tensor(self, value, dtype=None):
+                return convert_to_tensor(value, dtype=dtype, sparse=False)
+
+            # Overload native accessor.
+            def __jax_array__(self):
+                return self.value
+
+            @property
+            def value(self):
+                if not hasattr(self, "raw_value"):
+                    if not hasattr(self, "_value") or self._value is None:
+                        if self._initializer is not None:
+                            initial_value = self._initializer(
+                                self._shape, dtype=self._dtype
+                            )
+                            return self._maybe_autocast(initial_value)
+                        else:
+                            raise AttributeError(
+                                "Variable is not properly initialized and has"
+                                " no initializer."
+                            )
+                    current_value = self._value
+                else:
+                    current_value = self.raw_value
+                    if (
+                        hasattr(self, "_var_metadata")
+                        and "on_get_value" in self._var_metadata
+                    ):
+                        current_value = self._var_metadata["on_get_value"](
+                            self, current_value
+                        )
+
+                if in_stateless_scope():
+                    scope = get_stateless_scope()
+                    stateless_value = scope.get_current_value(self)
+                    if stateless_value is not None:
+                        return self._maybe_autocast(stateless_value)
+
+                return self._maybe_autocast(current_value)
+
+            def __hash__(self):
```

Comment:
Per the discussion we had, we can't do that. It's not compatible with the semantics of `__eq__` that we have.

I think my preferred fix would be to change NNX to not use variables as keys, but instead use `id(variable)` in dictionaries.

Reply:
done